include/mbgl/gfx/uniform_block.hpp

```diff
@@ -0,0 +1,91 @@
+#pragma once
+
+#include <string>
+
+namespace mbgl {
+namespace gfx {
+
+class UniformBuffer;
+
+class UniformBlock {
+    // Can only be created by UniformBlockArray implementations
+protected:
+    UniformBlock(int index_, std::size_t size_)
+        : index(index_),
+          size(size_) {}
+    UniformBlock(const UniformBlock&) = default;
+    UniformBlock(UniformBlock&& other)
+        : index(other.index),
+          size(other.size) {}
+
+public:
+    virtual ~UniformBlock() = default;
+
+    int getIndex() const { return index; }
+
+    std::size_t getSize() const { return size; }
+
+    virtual void bindBuffer(const UniformBuffer& uniformBuffer) = 0;
+    virtual void unbindBuffer() = 0;
+
+protected:
+    UniformBlock& operator=(const UniformBlock&) = default;
+    UniformBlock& operator=(UniformBlock&& other) {
+        index = other.index;
+        size = other.size;
+        return *this;
+    }
+
+protected:
+    int index;
+    std::size_t size;
+};
+
+/// Stores a collection of uniform blocks by name
+class UniformBlockArray {
+public:
+    using UniformBlockMap = std::unordered_map<std::string, std::unique_ptr<UniformBlock>>;
+
+    UniformBlockArray(int initCapacity = 10);
+    UniformBlockArray(UniformBlockArray&&);
+    // Would need to use the virtual assignment operator
+    UniformBlockArray(const UniformBlockArray&) = delete;
+    virtual ~UniformBlockArray() = default;
+
+    /// Get map of elements.
+    const UniformBlockMap& getMap() const { return uniformBlockMap; }
+
+    /// Number of elements
+    std::size_t size() const { return uniformBlockMap.size(); }
+
+    /// Get an uniform block element.
+    /// Returns a pointer to the element on success, or null if the uniform block doesn't exists.
+    UniformBlock* get(const std::string& name) const;
```

Comment:
Raw pointers don't bother me all that much, but I can see they bother some.
Should the UniformBlock* be some other C++ structure on return?

Reply:
I think this is based on the attribute stuff I did.  It's a `unique_ptr` internally and we don't return ownership, only a temporary reference, but it's optional so it's a pointer.  The nicer semantics would be a `shared_ptr`, but there's no actual need for the refcounting.  Maybe an `optional<reference_wrapper<>>`?

Reply:
We could return a `const std::unique_ptr<>&` temporary which would better convey ownership.

Reply:
That feels more C++

Reply:
I need to return `nullptr`, not sure if this is possible for this return `const std::unique_ptr<>&`.

Reply:
works for me

Reply:
A more C++-y way, but you could return an optional holding a reference wrapper. Ex: https://godbolt.org/z/3xdrnE3dG

Reply:
Yes, maybe we should create a global one somewhere.

Reply:
So we should used `optional<reference_wrapper<>>` as also Tim mentioned?

Reply:
I'm fine with the pointer for now. The reference wrapper is more verbose but does help convey the returned reference is temporary.

Reply:
I'm in favor of `const std::unique_ptr<>&`

Reply:
I changed it to `const std::unique_ptr<>&`.